loader/db_test.go

```diff
@@ -0,0 +1,154 @@
+package loader
+
+import (
+	"bufio"
+	"os"
+	"strconv"
+
+	. "github.com/pingcap/check"
+
+	"github.com/pingcap/dm/dm/config"
+	tcontext "github.com/pingcap/dm/pkg/context"
+)
+
+var _ = Suite(&testdDbSuite{})
+
+type testdDbSuite struct {
+	cfg *config.SubTaskConfig
+}
+
+func (t *testdDbSuite) SetUpSuite(c *C) {
+	host := os.Getenv("MYSQL_HOST")
+	if host == "" {
+		host = "127.0.0.1"
+	}
+	port, _ := strconv.Atoi(os.Getenv("MYSQL_PORT"))
+	if port == 0 {
+		port = 3306
+	}
+	user := os.Getenv("MYSQL_USER")
+	if user == "" {
+		user = "root"
+	}
+	pswd := os.Getenv("MYSQL_PSWD")
+
+	t.cfg = &config.SubTaskConfig{
+		To: config.DBConfig{
+			Host:     host,
+			User:     user,
+			Password: pswd,
+			Port:     port,
+		},
+		MetaSchema: "test",
+	}
+	t.cfg.To.Adjust()
+}
+
+func (t *testdDbSuite) TestProcessErrSQL(c *C) {
+	tctx := tcontext.Background()
+
+	conn, err := createConn(t.cfg)
+	c.Assert(err, IsNil)
+	defer closeConn(conn)
+
+	txn, err := conn.db.Begin()
+	c.Assert(err, IsNil)
+
+	_, err = txn.Exec("USE test;")
+	c.Assert(err, IsNil)
+
+	sql := "CREATE TABLE `testTableRoger` (`a` varchar(255) DEFAULT NULL,`b` varchar(255) NOT NULL,`c` varchar(255) DEFAULT NULL,`d` int(11) DEFAULT NULL,PRIMARY KEY (`b`)) ENGINE=InnoDB DEFAULT CHARSET=latin1;"
+
+	_, err = txn.Exec(sql)
+	c.Assert(err, IsNil)
+
+	for i := 0; i < 2; i++ {
+		// when i = 0, need to determine the existence of duplicate data
+
+		sql = "INSERT IGNORE INTO `testTableRoger` VALUES\n('111','aaa','!!!',1),\n('222','bbb','###',2),\n('333','ccc','$$$',3);"
+
+		_, err = txn.Exec(sql)
+		c.Assert(err, IsNil)
+	}
+
+	processErrSQL(tctx, conn, txn, sql)
+
+	sql = "DROP TABLE `testTableRoger`;"
+
+	_, err = txn.Exec(sql)
+	c.Assert(err, IsNil)
+
+	err = txn.Commit()
+	c.Assert(err, IsNil)
+}
+
+func (t *testdDbSuite) TestGetDupEntry(c *C) {
+	mes := "Duplicate entry 'aaa' for key 'PRIMARY'"
+
+	data, key, err := getDupEntry(mes)
+	c.Assert(data, Equals, "aaa")
+	c.Assert(key, Equals, "'PRIMARY'")
+	c.Assert(err, IsNil)
+
+	mes = "Duplicate entry 'aaa' key 'PRIMARY'"
+
+	data, key, err = getDupEntry(mes)
+	c.Assert(data, Equals, "")
+	c.Assert(key, Equals, "")
+	c.Assert(err, NotNil)
+}
+
+func (t *testdDbSuite) TestFindDataPair(c *C) {
+	target := "INSERT IGNORE INTO `testTableRoger` VALUES\n('111','aaa','!!!',1),\n('222','bbb','###',2),\n('333','ccc','$$$',3);"
+	dupData := "aaa"
+	pos := 2
+
+	data, err := findDataPair(target, dupData, pos)
+	c.Assert(data, Equals, "('111','aaa','!!!',1)")
+	c.Assert(err, IsNil)
+
+	pos = 3
+	data, err = findDataPair(target, dupData, pos)
+	c.Assert(data, Equals, "")
+	c.Assert(err, NotNil)
+}
+
+func (t *testdDbSuite) TestDupEntryLog(c *C) {
+	path := "./dupDataFile.txt"
```

Comment:
we can use c.MkDir() to create a test dir, and the file is under this dir. this dir will removed automatic